lib/rubocop/cop/style/hash_transform_methods.rb

```diff
@@ -0,0 +1,222 @@
+          (send
+            (const _ :Hash)
+            :[]
+            (block
+              ({send csend} _ :map)
+              (args
+                (arg $_)
+                (arg $_))
+              (array $_ $_)))
+        PATTERN
+
+        MAP_TO_H_PATTERN = NodePattern.new(<<~PATTERN)
+          ({send csend}
+            (block
+              ({send csend} _ :map)
+              (args
+                (arg $_)
+                (arg $_))
+              (array $_ $_))
+            :to_h)
+        PATTERN
+
+        VALS_MSG = 'Use `transform_values` to transform just the values of a ' \
+                   'hash, rather than `each_with_object` or `map`'
+        KEYS_MSG = 'Use `transform_keys` to transform just the keys of a ' \
+                   'hash, rather than `each_with_object` or `map`'
+
+        def on_block(node)
+          match = EACH_WITH_OBJECT_PATTERN.match(node)
+          return unless match
+
+          k_arg, v_arg, hash_arg, hash_body, k_body, v_body = *match
+          return unless hash_arg == hash_body
+
+          handle_possible_offense(node, k_arg, v_arg, k_body, v_body)
+        end
+
+        def on_send(node)
+          match = (
+            HASH_BRACKETS_MAP_PATTERN.match(node) ||
+            MAP_TO_H_PATTERN.match(node)
+          )
+          handle_possible_offense(node, *match) if match
+        end
+
+        def on_csend(node)
+          match = MAP_TO_H_PATTERN.match(node)
+          handle_possible_offense(node, *match) if match
+        end
+
+        def autocorrect(node)
+          lambda do |corrector|
+            correction = preprocess_by_match_type_for_autocorrect(
+              node,
+              corrector
+            )
+            correction.apply(corrector)
+          end
+        end
+
+        private
+
+        def handle_possible_offense(node, k_arg, v_arg, k_body, v_body)
+          unchanged_key = unchanged_key_or_val?(k_body, k_arg)
+          unchanged_val = unchanged_key_or_val?(v_body, v_arg)
+
+          # If both key and value are changed, this block can't be replaced.
+          # If neither is changed, this is either a false positive (receiver
+          # isn't a hash) or very weird; either way ignore it.
+          return if unchanged_key == unchanged_val
+
+          # Can't use `transform_values` if value transformation uses key,
+          # or `transform_keys` if key transformation uses value
+          if unchanged_key && !v_body.descendants.include?(k_body)
+            add_offense(node, message: VALS_MSG)
+          elsif unchanged_val && !k_body.descendants.include?(v_body)
+            add_offense(node, message: KEYS_MSG)
+          end
+        end
+
+        def unchanged_key_or_val?(body, arg)
+          body.lvar_type? && body.children == [arg]
+        end
+
+        def preprocess_by_match_type_for_autocorrect(node, corrector) # rubocop:disable Metrics/AbcSize
+          if (match = EACH_WITH_OBJECT_PATTERN.match(node))
```

Comment:
Seems to me you can use `switch` here.

Reply:
`Style/EmptyCaseCondition` prevents me from doing that here, if I understand the suggestion correctly.